cpp/src/io/utilities/column_utils.cuh

```diff
@@ -67,13 +67,17 @@ rmm::device_uvector<column_device_view> create_leaf_column_device_views(
       size_type index) mutable {
       col_desc[index].parent_column = parent_col_view.begin() + index;
       column_device_view col        = parent_col_view.column(index);
-      if (col_desc[index].stats_dtype != dtype_byte_array) {
-        // traverse till leaf column
-        while (col.type().id() == type_id::LIST or col.type().id() == type_id::STRUCT) {
-          col = (col.type().id() == type_id::LIST)
-                  ? col.child(lists_column_view::child_column_index)
-                  : col.child(0);
+      // traverse till leaf column
+      while (col.type().id() == type_id::LIST or col.type().id() == type_id::STRUCT) {
+        auto child = (col.type().id() == type_id::LIST)
+                       ? col.child(lists_column_view::child_column_index)
+                       : col.child(0);
+        // stop early if writing a byte array, it needs to be a list<int8> not the int8 column
+        if (col_desc[index].stats_dtype == dtype_byte_array &&
+            (child.type().id() == type_id::INT8 || child.type().id() == type_id::UINT8)) {
+          break;
         }
+        col = child;
```

Comment:
In case of nested struct, this will travel down to `child(0)` then `child(0).child(0)` then `child(0).child(0).child(0)` etc, not scanning all the children. Is this correct?

Reply:
Consider a column that is `list<list<list<int8>>>` in the case of a binary write:

The original code would march to the leaf column and return the `int8` column. The code as it is now will grab the top `list<list,list<int8>>>` and return that. The change marches until the last list and returns `list<int8>` as the leaf column. This is the correct method, as later in the code we reference that leaf column and expect to be able to get a `byte_array_view` out of it. This only works if the column we ask that about has the data child.

Reply:
Yes but that's the case for nested lists. I asked the case for nested structs in the code:
```
...or col.type().id() == type_id::STRUCT
```

Reply:
This is a `parquet_column_view` thing as shown [here](https://github.com/rapidsai/cudf/blob/9c22da57ee9919d89c59f321226e4438322ea073/cpp/src/io/parquet/writer_impl.cu#L719). A struct is ripped apart so that only one child exists per column. This is a parquet writing thing and relates to how parquet stores the data for structs.